### PR TITLE
Add 'await' operator to GDScript

### PR DESCRIPTION
**Pull request by guykayjay**
## Summary

This PR adds an `await` operator to GDScript, as a follow-up to the discussion from #30328.

## Motivation

Calling `yield()` with a yielding function call is clunky, especially when the function doesn't always yield in the course of its execution:

```gdscript
var x = could_yield()
if x is GDScriptFunctionState:
    x = yield(x, "completed")
```

This PR introduces an `await` operator to clear away this friction.

```gdscript
var x = await could_yield()
```

## Details

Formally speaking, the expression `await x` is equivalent to `yield(x, "completed") if x is GDScriptFunctionState else x`, except that `x` is only evaluated once. `await` is a unary operator with the highest precedence level, so it can be used deep within other expressions.

If the operand to `await` is not a `GDScriptFunctionState`, the operator has no effect (but a warning is emitted when using `await` is certainly redundant). This way, you can safely get the logical result from function calls that may not yield.

## Benefits

This operator would have a couple of benefits:
- Saves a lot of typing for one of the most common use cases of GDScript coroutines
- Accomplishes what is otherwise syntactically impossible without introducing a local variable
- Doesn't affect GDScript's bytecode format
- `await` is still a valid identifier for non-local variables, so APIs using this name will not break
- Fulfills a common request from the community (see #21039, #22947, #30328)

### CODE REVIEW

**Comment by Chaosus** [comment]
Not sure if this suitable for 3.2 - its on feature freeze stage. I think its more feature proposal than enchancement.

**Comment by Xrayez** [comment]
This is much needed, there was a feature proposal here which proposed having to add macros to GDScript, but the real motivation behind that was actually adding `await` keyword imo: #10948, #21308.

**Comment by Chaosus** [comment]
Well, as an exception this can be merged by @akien-mga

**Comment by rivok** [comment]
Update: Moved the `await` expression to the highest precedence level, as you would typically expect `await x is Y` to parse as `(await x) is Y`, not `await (x is Y)`.

**Comment by somnivore** [comment]
This is way better than my proposal (https://github.com/godotengine/godot/issues/21039), much rather have this just due to the implicit could-yield guard, let alone the nicer, simpler syntax.  @rivok do you know when the merge conflict will be resolved?  I'd like to test this out.

**Comment by rivok** [comment]
Rebased and resolved merge conflict

**Comment by asheraryam** [comment]
I'm doing a lot of networking and I really support this getting merged (or something similar) so we can have more straightforward asynchronous code.

**Comment by jabcross** [comment]
I just found this pull request, minutes after doing my own: #33224

The only difference seems to be the parenthesis.

File: modules/gdscript/gdscript_parser.cpp [hunk 1]
```diff
@@ -892,39 +933,6 @@ GDScriptParser::Node *GDScriptParser::_parse_expression(Node *p_parent, bool p_s
...
-			expression.push_back(e);
-			continue; //only exception, must continue...
-
-			/*
-			Node *subexpr=_parse_expression(op,p_static);
-			if (!subexpr)
-				return NULL;
-			op->arguments.push_back(subexpr);
-			expr=op;*/
-
```
Comment (rivok) [inline modules/gdscript/gdscript_parser.cpp]: Prefix operator parsing was moved up to share logic with `await`, which is also a prefix operator and needs to be parsed earlier than function calls. This change has no impact on the existing prefix operators.

**Comment by nathanwfranke** [comment]
I would personally prefer it to look like a function instead of a space-separated keyword. Currently the order is not clear and can lead to confusion or code like `(await myfunc()).otherfunc()` when it would be cleaner to say `await(myfunc()).otherfunc()`

**Comment by rivok** [comment]
@nathanwfranke Hmm, interesting, I hadn't thought of that. We should get more opinions on this. Personally, I like how it already is. In most cases I find the prefix style easier to read and type, and it's syntactically consistent with the `await` expressions of [Python](https://docs.python.org/3/reference/expressions.html#await), [C#](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/await), [JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await), and soon [C++](https://en.cppreference.com/w/cpp/language/coroutines#co_await).

I implemented the parenthesized version locally to see if we could parse both styles by greedily parsing `await(expr)` before `await expr`, but there are some weird cases (e.g. `await (x as T).y`) that bother me. That particular example could probably be "fixed" with a parser warning though.

**Comment by Calinou** [comment]
I'd prefer allowing only one syntax, like we now do with `assert()` to support multiple arguments. While it's unlikely `await()` will have to support multiple arguments, I think it's better to be consistent and encourage writing code that's easily refactorable.

**Comment by regakakobigman** [comment]
Love this proposal. I need to `yield(foo(), "completed")` a lot in my projects, and having to check `if value is GDScriptFunctionState` makes static typing impossible. This proposal solves both of those issues. I was a proponent of @somnivore's proposal to make `yield(foo())` an alias for `yield(foo(), "completed")`, but I like the dedicated `await` keyword and the additional functionality it provides it provides over plain `yield()`.

I personally prefer `await foo()`, but I agree with Calinou that `await(foo())` is more consistent. I also think @rivok's edge case is important enough that using `await foo()` is more trouble than it's worth.

**Comment by skison** [comment]
As much as I like the `await` keyword in other languages, I agree that having `await()` would seem more consistent, and I think I would actually prefer it over just the keyword.

But really, either implementation would make me happy!

**Comment by NathanLovato** [comment]
Same as above, I'm in favor of consistent syntax in GDScript, so using parentheses.

Recent versions of Godot removed the old syntax `assert condition` etc. in favor of using parentheses so it'd be unintuitive for new keywords to have a different syntax.

And the linter is there to tell you that you need parentheses with e.g. `assert` now.

**Comment by rivok** [comment]
Thanks for the feedback, everyone. It looks like the majority prefers function-style `await()`.

My question now is whether we need to introduce a new keyword at all? Originally, I imagined `await` would be seen as syntactic sugar, not a builtin function, so I thought making it look like other languages' `await` would make sense. If we aren't going in that direction, it might be cleaner just to use `yield()` for this, because `yield()` doesn't have a form that only accepts one argument.

This has been proposed before (by @somnivore #21039), but the key difference here would be the safe type check.

**Comment by jabcross** [comment]
I've made a proposal with the functionality I'd like `yield(x)` with a single parameter to have. As such, I'd rather keep `await` for this JavaScript `async/await` - like functionality, and `yield(x) `to have Python-like `yield` functionality,

https://github.com/godotengine/godot-proposals/issues/453

**Comment by eduardonunesp** [comment]
Any chance to have this merged on  `3.2.2`

**Comment by Calinou** [comment]
@eduardonunesp I don't think so, as 3.2.2 is very close to being released. Moreover, there are merge conflicts that need to be solved.

File: modules/gdscript/gdscript_parser.h [hunk 2]
```diff
@@ -209,7 +209,7 @@ class GDScriptParser {
 
 		bool _static;
 		MultiplayerAPI::RPCMode rpc_mode;
-		bool has_yield;
+		bool has_yield_or_await;
```
Comment (Xrayez) [inline modules/gdscript/gdscript_parser.h:212]:
```suggestion
		bool has_coroutine;
```
Just a small suggestion.

Comment (rivok) [reply modules/gdscript/gdscript_parser.h:212]: I like that. I'll have to change that over.

**Comment by rivok** [comment]
Nice to see renewed interest in this concept. When I get some time in the coming days I'll rebase off of 4.0 and implement the syntax change.

**Comment by jabcross** [comment]
I think @vnen was already planning to add it to GDScript 2.0.

**Comment by aaronfranke** [comment]
@vnen is currrently working on rewriting GDScript, so this PR would most likely have to be rewritten from scratch for it to be merged. Therefore, I'm closing this and adding "salvageable".

Of course, this feature is highly desired and will almost certainly make it in eventually. When this PR  was made, the plan to rewrite GDScript did not exist yet, and things have changed such that this isn't good as-is. Once the GDScript rewrite is merged, this feature can be redone and merged.

**Comment by KoBeWi** [comment]
AFAIK the `await` keyword has already been implemented in the not-yet-published GDScript rewrite, but its behavior might be different from the one proposed here.

**Comment by asheraryam** [comment]
What KoBeWi said https://twitter.com/vnen/status/1264621967526813696

**Comment by Mickeon** [comment]
Hasn't this been fully salvaged at this point, 2 years later?

**Comment by rivok** [comment]
> Hasn't this been fully salvaged at this point, 2 years later?

`await` made it into Godot 4 but it's implemented differently there. This PR was for 3.x which will definitely not be getting `await`.

**Pull request closed without merging**